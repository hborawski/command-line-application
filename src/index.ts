import commandLineArgs from 'command-line-args';
import commandLineUsage from 'command-line-usage';
import removeMarkdown from 'remove-markdown';

export type Option = commandLineUsage.OptionDefinition;

export type Example = {
  desc: string;
  example: string;
};

export type Section = commandLineUsage.Section & {
  code?: boolean;
};

export type Command = {
  name: string;
  description: string;
  options?: Option[];
  require?: string[];
  examples?: (string | Example)[];
  /** What group to render the command in a MultiCommand */
  group?: string;
  footer?: Section[] | Section | string;
};

export interface MultiCommand {
  name: string;
  logo?: string;
  description: string;
  options?: Option[];
  commands: (Command | MultiCommand)[];
  footer?: Section | string;
}

const help: Option = {
  name: 'help',
  alias: 'h',
  description: 'Display the help output',
  type: Boolean,
  group: 'global'
};

const globalOptions = [help];

const arrayify = <T>(x: T | T[]): T[] => (Array.isArray(x) ? x : [x]);
const hasGlobal = (options: Option[]) =>
  Boolean(options.find(option => option.group === 'global'));

function styleTypes(command: Command, option: Option) {
  const isRequired = command.require && command.require.includes(option.name);

  if (isRequired && option.type === Number) {
    option.typeLabel = `{rgb(173, 216, 230) {underline ${option.typeLabel ||
      'number'}}} [{rgb(254,91,92) required}]`;
  } else if (option.type === Number) {
    option.typeLabel = `{rgb(173, 216, 230) {underline ${option.typeLabel ||
      'number'}}}`;
  }

  if (isRequired && option.type === String) {
    option.typeLabel = `{rgb(173, 216, 230) {underline ${option.typeLabel ||
      'string'}}} [{rgb(254,91,92) required}]`;
  } else if (option.multiple && option.type === String) {
    option.typeLabel = `{rgb(173, 216, 230) {underline ${option.typeLabel ||
      'string[]'}}}`;
  } else if (option.type === String) {
    option.typeLabel = `{rgb(173, 216, 230) {underline ${option.typeLabel ||
      'string'}}}`;
  }
}

function addFooter(command: Command, sections: Section[]) {
  if (typeof command.footer === 'string') {
    sections.push({ content: command.footer });
  } else if (command.footer) {
    const footers = Array.isArray(command.footer)
      ? command.footer
      : [command.footer];

    footers.forEach(f => {
      // @ts-ignore
      if (typeof f.content === 'string') {
        console.log(
          'line',
          // @ts-ignore
          f.content
            .replace(/}/g, '\\}')
            .replace(/{/g, '\\{')
            .split('\n')
            // @ts-ignore
            .map(s => `  ${s}  `)
        );
      }

      sections.push({
        ...f,
        header: f.header
          ? removeMarkdown(f.header, { stripListLeaders: false })
          : undefined,
        content: !('content' in f)
          ? undefined
          : typeof f.content === 'string'
          ? removeMarkdown(
              f.code
                ? f.content
                    .replace(/}/g, '\\}')
                    .replace(/{/g, '\\{')
                    .split('\n')
                    .map(s => (s.includes('```') ? s : `  ${s}  `))
                    .join('\n')
                : f.content,
              {
                stripListLeaders: false
              }
            )
          : Array.isArray(f.content)
          ? f.content
          : undefined
      });
    });
  }
}

const printUsage = (command: Command) => {
  const options = command.options || [];
  const sections: Section[] = [
    {
      header: command.name,
      content: command.description
    }
  ];

  options.forEach(option => {
    styleTypes(command, option);
  });

  if (hasGlobal(options)) {
    sections.push(
      {
        header: 'Options',
        optionList: options
      },
      {
        header: 'Global Options',
        optionList: [...options, ...globalOptions],
        group: 'global'
      }
    );
  } else {
    sections.push({
      header: 'Options',
      optionList: [...options, ...globalOptions],
      group: ['_none', 'global']
    });
  }

  if (command.examples) {
    sections.push({
      header: 'Examples',
      content: command.examples
    });
  }

  addFooter(command, sections);

  console.log(commandLineUsage(sections));
  return;
};

const printRootUsage = (multi: MultiCommand) => {
  const subCommands =
    multi.commands.filter((c): c is Command => !('isMulti' in c)) || [];
  const rootOptions = multi.options || [];
  const options = [...rootOptions, ...globalOptions];
  const sections: Section[] = [];

  if (multi.logo) {
    sections.push({
      content: multi.logo,
      raw: true
    });
  }

  sections.push({
    header: multi.name,
    content: multi.description
  });

  sections.push({
    header: 'Synopsis',
    content: `$ ${multi.name} <command> <options>`
  });

  const groups = subCommands.reduce((all, command) => {
    if (command.group) {
      all.add(command.group);
    }

    return all;
  }, new Set<string>());

  groups.forEach(header => {
    const grouped = subCommands.filter(c => c.group === header) || [];

    if (grouped.length > 0) {
      sections.push({
        header,
        content: grouped.map(command => ({
          name: command.name,
          description: command.description
        }))
      });
    }
  });

  if (groups.size === 0) {
    sections.push({
      header: 'Commands',
      content: subCommands.map(command => ({
        name: command.name,
        description: command.description
      }))
    });
  }

  options.forEach(option => {
    styleTypes(multi, option);
  });

  sections.push({
    header: 'Global Options',
    optionList: options,
    group: ['_none', 'global']
  });

  addFooter(multi, sections);

  console.log(commandLineUsage(sections));
};

interface Options {
  argv?: string[];
  showHelp?: boolean;
}

const parseCommand = (
  command: Command,
  { argv, showHelp }: Options
): Record<string, any> | undefined => {
  const args = [...(command.options || []), help];
  const { global, ...rest } = commandLineArgs(args, {
    stopAtFirstUnknown: true,
    camelCase: true,
    argv
  });

  if (rest._unknown) {
    printUsage(command);
    console.log(
      `Found unknown flag${
        rest._unknown.length > 1 ? 's' : ''
      }: ${rest._unknown.join(', ')}`
    );
    return;
  }

  if (global.help && showHelp) {
    printUsage(command);
    return;
  }

  if (command.require) {
    const missing: string[] = [];

    command.require.forEach(r => {
      if (rest._all[r] === undefined) {
        missing.push(r);
      }
    });

    if (missing.length > 0) {
      const multiple = missing.length > 1;
      printUsage(command);
      console.log(
        `Missing required arg${multiple ? 's' : ''}: ${missing.join(', ')}`
      );
      process.exit(1);
    }
  }

  return { ...rest, ...rest._all, ...global };
};

export function app(
  command: Command | MultiCommand,
  { showHelp = true, argv }: Options = {}
):
  | ({ _command: string | string[] } & Record<string, any>)
  | Record<string, any>
  | undefined {
  const appOptions = { showHelp, argv };

  if (!('commands' in command)) {
    return parseCommand(command, appOptions);
  }

  const { global, _unknown } = commandLineArgs(globalOptions, {
    stopAtFirstUnknown: true,
    camelCase: true,
    argv
  });

  if (global.help && showHelp) {
    printRootUsage(command);
    return;
  }

  if (_unknown && _unknown.length > 0) {
    const rootOptions = command.options ? command.options || [] : [];
    const subCommand = command.commands.find((c): c is Command =>
      Boolean(c.name === _unknown[0])
    );

    if (subCommand) {
      const options = [...(subCommand.options || []), ...rootOptions];
      const parsed = app(
        { ...subCommand, options },
        { ...appOptions, argv: _unknown.slice(1) }
      );

      if (!parsed) {
        return;
      }

      return {
        ...parsed,
        _command:
          '_command' in parsed
            ? [subCommand.name, ...arrayify(parsed._command)]
            : subCommand.name
      };
    }

    printRootUsage(command);
    console.log(
      `Found unknown flag${_unknown.length > 1 ? 's' : ''}: ${_unknown.join(
        ', '
      )}`
    );

    return;
  } else {
    if (showHelp) {
      printRootUsage(command);
      console.log(`No sub-command provided to MultiCommand "${command.name}"`);
    }
    process.exit(1);
  }

  return global;
}
