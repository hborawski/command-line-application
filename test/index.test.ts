import { app, Command, MultiCommand } from '../src';

expect.addSnapshotSerializer({
  test: val => typeof val === 'string',
  print: (val: string) =>
    val
      .split('\n')
      .map(t => t.trimRight().replace('38;5;153m', '97m'))
      .join('\n'),
});

beforeEach(() => {
  process.argv = [];
  jest.clearAllMocks();
});

describe('single command app', () => {
  const echo: Command = {
    name: 'echo',
    description: 'Print a string to the terminal',
    examples: ['echo foo', 'echo "Intense message"'],
    options: [
      {
        name: 'value',
        type: String,
        defaultOption: true,
        description: 'The value to print',
      },
    ],
  };

  test('usage', () => {
    expect(app(echo, { argv: ['foo'] })).toEqual(
      expect.objectContaining({ value: 'foo' })
    );
  });

  test('help', () => {
    jest.spyOn(console, 'log').mockImplementationOnce(() => {});
    app(echo, { argv: ['--help'] });

    // @ts-ignore
    expect(console.log.mock.calls[0]).toMatchSnapshot();
  });
});

describe('multi command app', () => {
  const testCommand: Command = {
    name: 'test',
    description: 'test the project',
    examples: ['test --interactive'],
    options: [
      {
        name: 'interactive',
        type: Boolean,
        description: 'Run the application in interactive mode',
      },
    ],
  };
  const lintCommand: Command = {
    name: 'lint',
    description: 'lint the project',
    examples: ['lint --fix'],
    options: [
      {
        name: 'fix',
        type: Boolean,
        description: 'Run the application in fix mode',
      },
    ],
  };
  const scripts: MultiCommand = {
    name: 'scripts',
    description: 'My scripts package',
    commands: [lintCommand, testCommand],
    options: [
      {
        name: 'verbose',
        alias: 'v',
        description: 'Log a bunch of stuff',
        type: Boolean,
      },
    ],
  };

  test('help', () => {
    jest.spyOn(console, 'log').mockImplementationOnce(() => {});
    app(scripts, { argv: ['--help'] });

    // @ts-ignore
    expect(console.log.mock.calls[0]).toMatchSnapshot();
  });

  test('should exit when no sub command is provided', () => {
    // @ts-ignore
    jest.spyOn(process, 'exit').mockImplementationOnce(() => {});
    jest.spyOn(console, 'log').mockImplementationOnce(() => {});
    jest.spyOn(console, 'log').mockImplementationOnce(() => {});
    app(scripts);

    // @ts-ignore
    expect(process.exit).toHaveBeenCalledWith(1);
  });

  test('use lint', () => {
    expect(app(scripts, { argv: ['lint', '--fix'] })).toEqual(
      expect.objectContaining({ _command: 'lint', fix: true })
    );
  });

  test('use test', () => {
    expect(app(scripts, { argv: ['test'] })).toEqual(
      expect.objectContaining({ _command: 'test' })
    );
  });
});

describe('multi command app w/multi commands', () => {
  const testCommand: Command = {
    name: 'test',
    description: 'test the project',
    examples: ['test --interactive'],
    options: [
      {
        name: 'interactive',
        type: Boolean,
        description: 'Run the application in interactive mode',
      },
    ],
  };
  const createCommand: MultiCommand = {
    name: 'create',
    description: 'create a new project',
    // examples: ['create component', 'create project'],
    options: [
      {
        name: 'name',
        type: String,
        description: 'The name of the thing to create',
      },
    ],
    commands: [
      {
        name: 'component',
        description: 'create a component',
        examples: ['create component'],
      },
      {
        name: 'project',
        description: 'create a project',
        examples: ['create project'],
        options: [
          {
            name: 'cwd',
            type: Boolean,
            description: 'Use the current working directory',
          },
        ],
      },
    ],
  };
  const scripts: MultiCommand = {
    name: 'scripts',
    description: 'My scripts package',
    commands: [createCommand, testCommand],
    options: [
      {
        name: 'verbose',
        alias: 'v',
        description: 'Log a bunch of stuff',
        type: Boolean,
      },
    ],
  };

  test('sub-multi-command help', () => {
    jest.spyOn(console, 'log').mockImplementationOnce(() => {});
    app(scripts, { argv: ['create', '--help'] });

    // @ts-ignore
    expect(console.log.mock.calls[0]).toMatchSnapshot();
  });

  test('sub-multi-command sub-command help', () => {
    jest.spyOn(console, 'log').mockImplementationOnce(() => {});
    app(scripts, { argv: ['create', 'project', '--help'] });

    // @ts-ignore
    expect(console.log.mock.calls[0]).toMatchSnapshot();
  });

  test('sub-multi-command sub-command usage', () => {
    expect(
      app(scripts, { argv: ['create', 'project', '--cwd', '--name', 'foo'] })
    ).toEqual(
      expect.objectContaining({
        _command: ['create', 'project'],
        cwd: true,
        name: 'foo',
      })
    );
  });
});
