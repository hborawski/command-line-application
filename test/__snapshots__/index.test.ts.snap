// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multi command app errors on unknown flag 1`] = `
Array [
  "
scripts

  My scripts package 

Synopsis

  $ scripts <command> <options> 

Commands

  lint   lint the project 
  test   test the project 

Global Options

  -v, --verbose    Log a bunch of stuff    
  -h, --help       Display the help output 
",
]
`;

exports[`multi command app errors on unknown flag 2`] = `
Array [
  "Found unknown flag: --error",
]
`;

exports[`multi command app errors on unknown flag for sub command 1`] = `
Array [
  "
lint

  lint the project 

Options

  --fix            Run the application in fix mode 
  -v, --verbose    Log a bunch of stuff            
  -h, --help       Display the help output         

Examples

  lint --fix 
",
]
`;

exports[`multi command app errors on unknown flag for sub command 2`] = `
Array [
  "Found unknown flag: --error",
]
`;

exports[`multi command app help 1`] = `
Array [
  "
scripts

  My scripts package 

Synopsis

  $ scripts <command> <options> 

Commands

  lint   lint the project 
  test   test the project 

Global Options

  -v, --verbose    Log a bunch of stuff    
  -h, --help       Display the help output 
",
]
`;

exports[`multi command app suggests misspelled command 1`] = `
Array [
  "
scripts

  My scripts package 

Synopsis

  $ scripts <command> <options> 

Commands

  lint   lint the project 
  test   test the project 

Global Options

  -v, --verbose    Log a bunch of stuff    
  -h, --help       Display the help output 
",
]
`;

exports[`multi command app suggests misspelled command 2`] = `
Array [
  "Found unknown command \\"flint\\", did you mean \\"lint\\"?",
]
`;

exports[`multi command app w/multi commands sub-multi-command help 1`] = `
Array [
  "
create

  create a new project 

Synopsis

  $ create <command> <options> 

Commands

  component   create a component 
  project     create a project   

Global Options

  --name string    The name of the thing to create 
  -v, --verbose    Log a bunch of stuff            
  -h, --help       Display the help output         
",
]
`;

exports[`multi command app w/multi commands sub-multi-command sub-command help 1`] = `
Array [
  "
project

  create a project 

Options

  --cwd            Use the current working directory 
  --name string    The name of the thing to create   
  -v, --verbose    Log a bunch of stuff              
  -h, --help       Display the help output           

Examples

  create project 
",
]
`;

exports[`should display code 1`] = `
"
echo

  Print a string to the terminal 

Options

  -h, --help    Display the help output 

Examples

  echo foo               
  echo \\"Intense message\\" 

  |                                                                             
                                                                                
  |         {                                                                   
  |           \\"create\\": {                                                       
  |             \\"package\\": {                                                    
  |               \\"templates\\": [                                                
  |                 {                                                           
  |                   \\"name\\": \\"custom-js\\",                                      
  |                   \\"url\\": \\"https://github.intuit.com/me/my-template\\",        
  |                   \\"description\\": \\"A custom package template\\",               
  |                   \\"sha\\": \\"4b9c7b627307380fb31acae059f7c095d0c626b8\\"         
  |                 }                                                           
  |               ]                                                             
  |             }                                                               
  |           }                                                                 
  |         }                                                                   
                                                                                
"
`;

exports[`single command app errors on unknown flag 1`] = `
Array [
  "
echo

  Print a string to the terminal 

Options

  --value string   The value to print      
  -h, --help       Display the help output 

Examples

  echo foo               
  echo \\"Intense message\\" 
",
]
`;

exports[`single command app errors on unknown flag 2`] = `
Array [
  "Found unknown flag: --error",
]
`;

exports[`single command app help 1`] = `
Array [
  "
echo

  Print a string to the terminal 

Options

  --value string   The value to print      
  -h, --help       Display the help output 

Examples

  echo foo               
  echo \\"Intense message\\" 
",
]
`;

exports[`single command app suggests typo correction 1`] = `
Array [
  "
echo

  Print a string to the terminal 

Options

  --value string   The value to print      
  -h, --help       Display the help output 

Examples

  echo foo               
  echo \\"Intense message\\" 
",
]
`;

exports[`single command app suggests typo correction 2`] = `
Array [
  "Found unknown flag \\"--vakue\\", did you mean \\"value\\"?",
]
`;
